src/topology/gluing.lean

```diff
@@ -0,0 +1,259 @@
+  refl
+end
+
+lemma ι_jointly_surjective (x : 𝖣 .glued) : ∃ i (y : D.U i), 𝖣 .ι i y = x :=
+𝖣 .ι_jointly_surjective (forget Top) x
+
+/--
+An equivalence relation on `Σ i, D.U i` that holds iff `𝖣 .ι i x = 𝖣 .ι j y`.
+See `Top.glue_data.ι_eq_iff_rel`.
+-/
+def rel (a b : Σ i, ((D.U i : Top) : Type*)) : Prop :=
+  a = b ∨ ∃ (x : D.V (a.1, b.1)) , D.f _ _ x = a.2 ∧ D.f _ _ (D.t _ _ x) = b.2
+
+lemma rel_equiv : equivalence D.rel :=
+⟨ λ x, or.inl (refl x),
+  begin
+    rintros a b (⟨⟨⟩⟩|⟨x,e₁,e₂⟩),
+    exacts [or.inl rfl, or.inr ⟨D.t _ _ x, by simp [e₁, e₂]⟩]
+  end,
+  begin
+    rintros ⟨i,a⟩ ⟨j,b⟩ ⟨k,c⟩ (⟨⟨⟩⟩|⟨x,e₁,e₂⟩), exact id,
+    rintro (⟨⟨⟩⟩|⟨y,e₃,e₄⟩), exact or.inr ⟨x,e₁,e₂⟩,
+    let z := (pullback_iso_prod_subtype (D.f j i) (D.f j k)).inv ⟨⟨_,_⟩, e₂.trans e₃.symm⟩,
+    have eq₁ : (D.t j i) ((pullback.fst : _ ⟶ D.V _) z) = x := by simp,
+    have eq₂ : (pullback.snd : _ ⟶ D.V _) z = y := pullback_iso_prod_subtype_inv_snd_apply _ _ _,
+    clear_value z,
+    right,
+    use (pullback.fst : _ ⟶ D.V (i, k)) (D.t' _ _ _ z),
+    dsimp only at *,
+    substs e₁ e₃ e₄ eq₁ eq₂,
+    have h₁ : D.t' j i k ≫ pullback.fst ≫ D.f i k = pullback.fst ≫ D.t j i ≫ D.f i j,
+    { rw ← 𝖣 .t_fac_assoc, congr' 1, exact pullback.condition },
+    have h₂ : D.t' j i k ≫ pullback.fst ≫ D.t i k ≫ D.f k i =
+      pullback.snd ≫ D.t j k ≫ D.f k j,
+    { rw ← 𝖣 .t_fac_assoc,
+      apply @epi.left_cancellation _ _ _ _ (D.t' k j i),
+      rw [𝖣 .cocycle_assoc, 𝖣 .t_fac_assoc, 𝖣 .t_inv_assoc],
+      exact pullback.condition.symm },
+    exact ⟨continuous_map.congr_fun h₁ z, continuous_map.congr_fun h₂ z⟩
+  end⟩
+
+open category_theory.limits.walking_parallel_pair
+
+lemma eqv_gen_of_π_eq {x y : ∐ D.U} (h : 𝖣 .π x = 𝖣 .π y) :
+  eqv_gen (types.coequalizer_rel 𝖣 .diagram.fst_sigma_map 𝖣 .diagram.snd_sigma_map) x y :=
+begin
+  delta glue_data.π multicoequalizer.sigma_π at h,
+  simp_rw comp_app at h,
+  replace h := (Top.mono_iff_injective (multicoequalizer.iso_coequalizer 𝖣 .diagram).inv).mp _ h,
+  let diagram := parallel_pair 𝖣 .diagram.fst_sigma_map 𝖣 .diagram.snd_sigma_map ⋙ forget _,
+  have : colimit.ι diagram one x = colimit.ι diagram one y,
+  { rw ←ι_preserves_colimits_iso_hom,
+    simp [h] },
+  have :
+    (colimit.ι diagram _ ≫ colim.map _ ≫ (colimit.iso_colimit_cocone _).hom) _ =
+    (colimit.ι diagram _ ≫ colim.map _ ≫ (colimit.iso_colimit_cocone _).hom) _ :=
+    (congr_arg (colim.map (diagram_iso_parallel_pair diagram).hom
+    ≫ (colimit.iso_colimit_cocone (types.coequalizer_colimit _ _)).hom) this : _),
+  simp only [eq_to_hom_refl, types_comp_apply, colimit.ι_map_assoc,
+    diagram_iso_parallel_pair_hom_app, colimit.iso_colimit_cocone_ι_hom, types_id_apply] at this,
+  exact quot.eq.1 this,
+  apply_instance
+end
+
+lemma inv_image.equivalence {α : Sort u} {β : Sort v} (r : β → β → Prop) (f : α → β)
+  (h : equivalence r) : equivalence (inv_image r f) :=
+⟨λ _, h.1 _, λ _ _ x, h.2.1 x, inv_image.trans r f h.2.2⟩
```

Comment:
Where should this go?
Other `inv_image` lemmas are all in core lean.

Reply:
Maybe something like `logic/function/basic.lean`